File: ios/MenCare/Features/Admin/Manage/NotificationSettingsTab.swift
import SwiftUI
import Observation

/// Hours-before choices of the reminder at `index`: days for the first, hours for the others.
private func reminderOptions(_ index: Int) -> [Int] {
    index == 0 ? [24, 48] : [2, 4]
}

private func reminderLabel(_ hours: Int) -> String {
    if hours == 24 { return L("ntf_one_day_before") }
    return hours > 24 ? L("ntf_days_before", hours / 24) : L("ntf_hours_before", hours)
}

/// Anche questa scheda leggeva il repository dentro `body`. Adesso ha il suo
/// view model: le impostazioni arrivano dall'API (interruttori e promemoria,
/// due rotte diverse unite in un oggetto solo) e ogni modifica torna indietro.
@MainActor
@Observable
final class NotificationSettingsViewModel {

    private let admin: AdminRepository

    let state = LoadState()
    private(set) var settings = NotificationSettings(reminders: [])
    private(set) var saving = false
    var saveError: String?

    init(admin: AdminRepository) {
        self.admin = admin
    }

    func load() async {
        await state.run { settings = try await admin.notificationSettings() }
    }

    /// Ogni interruttore scrive subito: e' una schermata senza tasto "Salva",
    /// quindi l'esito va mostrato li' per li'.
    func update(_ transform: (inout NotificationSettings) -> Void) {
        var updated = settings
        transform(&updated)
        // Aggiornamento ottimista: l'interruttore si muove subito, e se il
        // server rifiuta si rimette com'era.
        let previous = settings
        settings = updated
        saving = true
        saveError = nil
        Task {
            switch await admin.updateNotificationSettings(updated) {
            case .success(let stored): settings = stored
            case .failure(let error):
                settings = previous
                saveError = error.displayMessage
            }
            saving = false
        }
    }
}

struct NotificationSettingsTab: View {
    @State var viewModel: NotificationSettingsViewModel

    var body: some View {
        Loadable(state: viewModel.state, retry: { Task { await viewModel.load() } }) {
            content(viewModel.settings)
        }
        .task { await viewModel.load() }
    }

    private func content(_ settings: NotificationSettings) -> some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                FormErrorBanner(message: viewModel.saveError)
                BrandSectionLabel(text: L("ntf_reminders"))
                remindersPanel(settings)

                BrandSectionLabel(text: L("ntf_types"))
                    .padding(.top, 4)
                typesPanel(settings)

                previewPanel
                    .padding(.top, 4)
            }
            .padding(.horizontal, 20)
            .padding(.top, 16)
            .padding(.bottom, 24)
            .readableWidth()
        }
    }

    private func update(_ transform: (inout NotificationSettings) -> Void) {
        viewModel.update(transform)
    }

    private func remindersPanel(_ settings: NotificationSettings) -> some View {
        VStack(spacing: 0) {
            ForEach(Array(settings.reminders.enumerated()), id: \.element.id) { index, rule in
                if index > 0 {
                    Rectangle().fill(Color.bone.opacity(0.1)).frame(height: 1)
                }
                VStack(spacing: 10) {
                    HStack {
                        Text(L("ntf_nth_reminder", index + 1))
                            .font(Typo.bodyLarge)
                            .foregroundStyle(Color.bone)
                        Spacer()
                        Button {
                            update { settings in
                                settings.reminders.removeAll { $0.id == rule.id }
                            }
                        } label: {
                            Image(systemName: "trash")
                                .font(.system(size: 15))
                                .foregroundStyle(Color.oliveWood)
                                .frame(width: 36, height: 36)
                                .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                        .accessibilityLabel(L("ntf_remove_reminder"))
                    }
                    let options = reminderOptions(index)
                    SegmentedTabs(
                        options: options.map(reminderLabel),
                        selectedIndex: options.firstIndex(of: rule.hoursBefore) ?? -1,
                        onSelect: { selectedIndex in
                            update { settings in
                                settings.reminders = settings.reminders.map {
                                    var updated = $0
                                    if $0.id == rule.id { updated.hoursBefore = options[selectedIndex] }
                                    return updated
                                }
                            }
                        }
                    )
                }
                .padding(.vertical, 12)
            }
            Button {
                update { settings in
                    // Parte da un'opzione che nessun altro promemoria usa ancora, se c'è.
                    let options = reminderOptions(settings.reminders.count)
                    let hours = options.first { h in !settings.reminders.contains { $0.hoursBefore == h } } ?? options[0]
                    settings.reminders.append(
                        ReminderRule(id: "rem_\(settings.reminders.count + 1)_\(Int(Date().timeIntervalSince1970 * 1000))", hoursBefore: hours)
                    )
                }
            } label: {
                HStack(spacing: 8) {
                    Image(systemName: "plus").font(.system(size: 13))
                    Text(L("ntf_add_reminder")).font(Typo.titleSmall)
                }
                .foregroundStyle(Color.bone)
                .frame(maxWidth: .infinity)
                .frame(height: 48)
                .overlay(RoundedRectangle(cornerRadius: 14).strokeBorder(Color.bone.opacity(0.25), lineWidth: 1))
            }
            .buttonStyle(.plain)
            .padding(.vertical, 10)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 6)
        .background(RoundedRectangle(cornerRadius: 18).fill(Color.ink))
    }

    private func typesPanel(_ settings: NotificationSettings) -> some View {
        VStack(spacing: 0) {
            toggleRow(L("ntf_confirmation"), hint: L("ntf_confirmation_hint"), value: settings.bookingConfirmation) { on in
                update { $0.bookingConfirmation = on }
            }
            Rectangle().fill(Color.stoneBorder).frame(height: 1)
            toggleRow(L("ntf_cancellation"), hint: nil, value: settings.cancellationAlert) { on in
                update { $0.cancellationAlert = on }
            }
            Rectangle().fill(Color.stoneBorder).frame(height: 1)
            toggleRow(L("ntf_late"), hint: L("ntf_late_hint"), value: settings.lateOperatorAlert) { on in
                update { $0.lateOperatorAlert = on }
            }
            Rectangle().fill(Color.stoneBorder).frame(height: 1)
            toggleRow(L("ntf_promo_empty"), hint: L("ntf_promo_empty_hint"), value: settings.emptyDayPromos) { on in
                update { $0.emptyDayPromos = on }
            }
        }
        .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }

    private func toggleRow(_ title: String, hint: String?, value: Bool, onChange: @escaping (Bool) -> Void) -> some View {
        HStack {
            VStack(alignment: .leading, spacing: 0) {
                Text(title).font(Typo.bodyLarge).foregroundStyle(Color.ink)
                if let hint {
                    Text(hint).font(Typo.jost(12)).foregroundStyle(Color.textMuted)
                }
            }
            Spacer()
            BrandSwitch(isOn: Binding(get: { value }, set: onChange))
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
    }

    private var previewPanel: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text(L("ntf_preview").uppercased())
                .font(Typo.jost(11, weight: .medium))
                .kerning(1.3)
                .foregroundStyle(Color.oliveWood)
            HStack(spacing: 11) {
                RoundedRectangle(cornerRadius: 10)
                    .fill(Color.ink)
                    .frame(width: 34, height: 34)
                VStack(alignment: .leading, spacing: 0) {
                    Text(L("ntf_preview_sender"))
                        .font(Typo.titleSmall)
                        .foregroundStyle(Color.ink)
                    Text(L("ntf_preview_body"))
                        .font(Typo.jost(12))
                        .foregroundStyle(Color.ink)
                }
                Spacer()
                Text(L("ntf_preview_now"))
                    .font(Typo.jost(11))
                    .foregroundStyle(Color.textMuted)
            }
        }
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .overlay(RoundedRectangle(cornerRadius: 16).strokeBorder(Color.oliveWood, lineWidth: 1.5))
    }
}

File: android/feature/admin/src/main/kotlin/com/devora/mencare/feature/admin/manage/NotificationSettingsTab.kt
package com.devora.mencare.feature.admin.manage

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material.icons.outlined.Delete
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.BrandSwitch
import com.devora.mencare.core.designsystem.component.SegmentedTabs
import com.devora.mencare.core.designsystem.component.readableWidth
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.OliveWood
import com.devora.mencare.core.designsystem.theme.Radii
import com.devora.mencare.core.designsystem.theme.StoneBorder
import com.devora.mencare.core.designsystem.theme.TextMuted
import com.devora.mencare.feature.admin.R

@Composable
internal fun NotificationSettingsTab(viewModel: NotificationSettingsViewModel = hiltViewModel()) {
    val settings by viewModel.settings.collectAsStateWithLifecycle()

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .readableWidth()
            .padding(horizontal = 20.dp)
            .padding(top = 16.dp, bottom = 24.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        BrandSectionLabel(stringResource(R.string.ntf_reminders))
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .clip(Radii.Md)
                .background(Bone)
                .border(1.5.dp, StoneBorder, Radii.Md)
                .padding(horizontal = 16.dp, vertical = 6.dp),
        ) {
            settings.reminders.forEachIndexed { index, rule ->
                if (index > 0) HorizontalDivider(color = StoneBorder)
                Column(modifier = Modifier.fillMaxWidth().padding(vertical = 12.dp)) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Text(
                            stringResource(R.string.ntf_nth_reminder, index + 1),
                            style = MaterialTheme.typography.bodyLarge,
                            color = Ink,
                            modifier = Modifier.weight(1f),
                        )
                        IconButton(onClick = { viewModel.removeReminder(rule.id) }) {
                            Icon(
                                Icons.Outlined.Delete,
                                contentDescription = stringResource(R.string.ntf_remove_reminder),
                                tint = OliveWood,
                            )
                        }
                    }
                    Spacer(Modifier.height(4.dp))
                    val options = reminderOptions(index)
                    SegmentedTabs(
                        options = options.map { hours -> reminderLabel(hours) },
                        selectedIndex = options.indexOf(rule.hoursBefore),
                        onSelect = { viewModel.setReminderHours(rule.id, options[it]) },
                        onDark = false,
                    )
                }
            }
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 10.dp)
                    .height(48.dp)
                    .clip(Radii.Md)
                    .border(1.5.dp, Ink, Radii.Md)
                    .clickable(onClick = viewModel::addReminder),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(Icons.Outlined.Add, contentDescription = null, tint = Ink, modifier = Modifier.size(16.dp))
                Text(
                    stringResource(R.string.ntf_add_reminder),
                    style = MaterialTheme.typography.titleSmall,
                    color = Ink,
                    modifier = Modifier.padding(start = 8.dp),
                )
            }
        }

        Spacer(Modifier.height(4.dp))
        BrandSectionLabel(stringResource(R.string.ntf_types))
        Column(
            modifier = Modifier.fillMaxWidth().clip(Radii.Md).background(Bone).border(1.5.dp, StoneBorder, Radii.Md),
        ) {
            ToggleRow(
                title = stringResource(R.string.ntf_confirmation),
                hint = stringResource(R.string.ntf_confirmation_hint),
                checked = settings.bookingConfirmation,
            ) { checked -> viewModel.update { it.copy(bookingConfirmation = checked) } }
            HorizontalDivider(color = StoneBorder)
            ToggleRow(
                title = stringResource(R.string.ntf_cancellation),
                hint = null,
                checked = settings.cancellationAlert,
            ) { checked -> viewModel.update { it.copy(cancellationAlert = checked) } }
            HorizontalDivider(color = StoneBorder)
            ToggleRow(
                title = stringResource(R.string.ntf_late),
                hint = stringResource(R.string.ntf_late_hint),
                checked = settings.lateOperatorAlert,
            ) { checked -> viewModel.update { it.copy(lateOperatorAlert = checked) } }
            HorizontalDivider(color = StoneBorder)
            ToggleRow(
                title = stringResource(R.string.ntf_promo_empty),
                hint = stringResource(R.string.ntf_promo_empty_hint),
                checked = settings.emptyDayPromos,
            ) { checked -> viewModel.update { it.copy(emptyDayPromos = checked) } }
        }

        Spacer(Modifier.height(4.dp))
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(16.dp))
                .border(1.5.dp, OliveWood, RoundedCornerShape(16.dp))
                .padding(14.dp),
        ) {
            Text(
                stringResource(R.string.ntf_preview).uppercase(),
                style = MaterialTheme.typography.labelMedium.copy(fontSize = 11.sp, letterSpacing = 0.14.em),
                color = OliveWood,
            )
            Spacer(Modifier.height(10.dp))
            Row(verticalAlignment = Alignment.CenterVertically) {
                Box(
                    modifier = Modifier.size(34.dp).clip(RoundedCornerShape(10.dp)).background(Ink),
                )
                Column(Modifier.weight(1f).padding(start = 11.dp)) {
                    Text(
                        stringResource(R.string.ntf_preview_sender),
                        style = MaterialTheme.typography.titleSmall,
                        color = Ink,
                    )
                    Text(
                        stringResource(R.string.ntf_preview_body),
                        style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp, lineHeight = 18.sp),
                        color = Ink,
                    )
                }
                Text(
                    stringResource(R.string.ntf_preview_now),
                    style = MaterialTheme.typography.bodySmall.copy(fontSize = 11.sp),
                    color = TextMuted,
                )
            }
        }
    }
}

@Composable
private fun reminderLabel(hours: Int): String = when {
    hours == 24 -> stringResource(R.string.ntf_one_day_before)
    hours > 24 -> stringResource(R.string.ntf_days_before, hours / 24)
    else -> stringResource(R.string.ntf_hours_before, hours)
}

@Composable
private fun ToggleRow(
    title: String,
    hint: String?,
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
) {
    Row(
        modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f)) {
            Text(title, style = MaterialTheme.typography.bodyLarge, color = Ink)
            if (hint != null) {
                Text(
                    hint,
                    style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                    color = TextMuted,
                )
            }
        }
        BrandSwitch(checked = checked, onCheckedChange = onCheckedChange)
    }
}
